src/dmd/expressionsem.d

```diff
@@ -6228,6 +6229,35 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 }
             }
         }
+        /* https://issues.dlang.org/show_bug.cgi?id=809
+         *
+         * If the address of a lazy variable is taken,
+         * the expression is rewritten so that the type
+         * of it is the delegate type. This means that
+         * the symbol is not going to represent a call
+         * to the delegate anymore, but rather, the
+         * actual symbol.
+         */
+
+        if (exp.e1.op == TOK.variable)
+        {
+            VarExp ve = cast(VarExp)exp.e1;
+            if (ve.var.storage_class & STC.lazy_)
+            {
+                exp.e1 = exp.e1.expressionSemantic(sc);
+                exp.e1 = resolveProperties(sc, exp.e1);
+                if (exp.e1.op == TOK.call)
+                {
+                    auto callExp = cast(CallExp)exp.e1;
+                    if (callExp.e1.type.ty == Tdelegate)
+                    {
+                        result = callExp.e1;
+                        return;
+                    }
+                }
+            }
+        }
+
         exp.e1 = exp.e1.toLvalue(sc, null);
```

Comment:
Is it ok that `exp.e1` here and following all of this could be mutated by lines 6247 and 6248 if it is a TOK.variable?

Reply:
It should be fine because it is only performing semantic analysis, thus enriching the AST. In order to be sure that you have a delegate there you need to do some transformations.